crates/matrix-sdk/src/notification_settings/mod.rs

```diff
@@ -12,3 +26,751 @@ pub enum RoomNotificationMode {
+
+        _ = rules.insert_room_rule(RuleKind::Override, &room_id, true).unwrap();
+        notification_settings.set_ruleset(&rules.ruleset).await;
+        assert_eq!(notification_settings.get_custom_rules_for_room(&room_id).await.len(), 2);
+    }
+
+    #[async_test]
+    async fn test_get_user_defined_room_notification_mode() {
+        let server = MockServer::start().await;
+        let client = logged_in_client(Some(server.uri())).await;
+        let room_id = get_test_room_id();
+
+        let notification_settings = client.notification_settings().await;
+        assert!(notification_settings
+            .get_user_defined_room_notification_mode(&room_id)
+            .await
+            .is_none());
+
+        let mut rules = notification_settings.rules().await;
+        // Set a notifying `Room` rule into the ruleset to be in `AllMessages`
+        _ = rules.insert_room_rule(RuleKind::Room, &room_id, true).unwrap();
+        notification_settings.set_ruleset(&rules.ruleset).await;
+        assert_eq!(
+            notification_settings.get_user_defined_room_notification_mode(&room_id).await.unwrap(),
+            RoomNotificationMode::AllMessages
+        );
+
+        // Set a mute `Room` rule into the ruleset to be in `MentionsAndKeywordsOnly`
+        _ = rules.insert_room_rule(RuleKind::Room, &room_id, false).unwrap();
+        notification_settings.set_ruleset(&rules.ruleset).await;
+        assert_eq!(
+            notification_settings.get_user_defined_room_notification_mode(&room_id).await.unwrap(),
+            RoomNotificationMode::MentionsAndKeywordsOnly
+        );
+
+        // Set a mute `Override` rule into the ruleset to be in `Mute`
+        _ = rules.insert_room_rule(RuleKind::Override, &room_id, false).unwrap();
+        notification_settings.set_ruleset(&rules.ruleset).await;
+        assert_eq!(
+            notification_settings.get_user_defined_room_notification_mode(&room_id).await.unwrap(),
+            RoomNotificationMode::Mute
+        );
+    }
+
+    #[async_test]
+    async fn test_get_default_room_notification_mode() {
+        let server = MockServer::start().await;
+        let client = logged_in_client(Some(server.uri())).await;
+
+        let notification_settings = client.notification_settings().await;
+        let mut rules = notification_settings.rules().await;
+
+        // notification_settings.get_default_room_notification_mode() should return the
+        // same values as rules.get_default_room_notification_mode()
+        rules
+            .ruleset
+            .set_actions(
+                RuleKind::Underride,
+                PredefinedUnderrideRuleId::RoomOneToOne,
+                vec![Action::Notify],
+            )
+            .unwrap();
+        notification_settings.set_ruleset(&rules.ruleset).await;
+        assert_eq!(
+            notification_settings.get_default_room_notification_mode(false, 2).await,
+            rules.get_default_room_notification_mode(false, 2)
```

Comment:
For this test, instead of comparing against the result of two functions, any chance we could put a raw value in there? (and duplicate the `assert_eq` if needs be)

In terms of testing, this doesn't tell us if the code is really working; it's just saying that these two functions are behaving the same way, but maybe they're behaving the same way in a manner that's incorrect/unexpected/unwanted.

Reply:
I've modified the test. I now test the expected result of the function instead of trying to test its implementation.